x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/action/InternalExecutePolicyAction.java

```diff
@@ -0,0 +1,170 @@
+            var node = selectNodeForPolicyExecution(clusterState.nodes());
+            if (clusterState.nodes().getLocalNode().equals(node) == false) {
+                var handler = new ActionListenerResponseHandler<>(actionListener, Response::new);
+                transportService.sendRequest(node, NAME, request, handler);
+                return;
+            }
+
+            // Can't use provided task, because in the case wait_for_completion=false then
+            // as soon as actionListener#onResponse is invoked then the provided task get unregistered and
+            // then there no way to see the policy execution in the list tasks or get task APIs.
+            var task = (ExecuteEnrichPolicyTask) taskManager.register("enrich", TASK_ACTION, new TaskAwareRequest() {
+
+                @Override
+                public void setParentTask(TaskId taskId) {
+                    request.setParentTask(taskId);
+                }
+
+                @Override
+                public TaskId getParentTask() {
+                    return request.getParentTask();
+                }
+
+                @Override
+                public Task createTask(long id, String type, String action, TaskId parentTaskId, Map<String, String> headers) {
+                    String description = "executing enrich policy [" + request.getName() + "]";
+                    return new ExecuteEnrichPolicyTask(id, type, action, description, parentTaskId, headers);
+                }
+            });
+
+            try {
+                ActionListener<ExecuteEnrichPolicyStatus> listener;
+                if (request.isWaitForCompletion()) {
+                    listener = ActionListener.wrap(result -> actionListener.onResponse(new Response(result)), actionListener::onFailure);
+                } else {
+                    listener = ActionListener.wrap(result -> {
+                        LOGGER.debug("successfully executed policy [{}]", request.getName());
+                        // taskManager.storeResult(task, new Response(result), ActionListener.wrap(() -> {/*noop*/}));
+                    }, e -> {
+                        LOGGER.error("failed to execute policy [" + request.getName() + "]", e);
+                        // taskManager.storeResult(task, e, ActionListener.wrap(() -> {/*noop*/}));
+                    });
+                }
+                policyExecutor.runPolicyLocally(task, request.getName(), ActionListener.wrap(result -> {
+                    taskManager.unregister(task);
+                    listener.onResponse(result);
+                }, e -> {
+                    taskManager.unregister(task);
+                    listener.onFailure(e);
+                }));
+
+                if (request.isWaitForCompletion() == false) {
+                    TaskId taskId = new TaskId(clusterState.nodes().getLocalNodeId(), task.getId());
+                    actionListener.onResponse(new Response(taskId));
```

Comment:
Good catch, I missed this. Yes, it does unlock the policy on the master node, which shouldn't happen.
The policy should only be unlocked here if wait_for_completing is set to `true`.

I think in the case if `wait_for_completion` is set to `false` then the response should be returned without unlocking the policy.  At the same time the get task action should be invoked with
that task id with `wait_for_completion=true` and if that returns then the policy should be unlocked.

Reply:
What do you think about something like this: c8822544aa758abb404414e85468d58abc2c3d49 ?

Reply:
That certainly could work!

A second thought: Since the master would have a listener waiting on the task to complete anyway, is there any compelling reason that the master node shouldn't handle the `wait_for_completion` feature itself? Making the internal policy action always wait for completion might simplify the code.

There's a bit more load put on the master in that second case (creating the task and cleaning it up), but that seems quite small compared to orchestrating the entire policy execution.

Reply:
> Since the master would have a listener waiting on the task to complete anyway, is there any compelling reason that the master node shouldn't handle the wait_for_completion feature itself? 

I think this is not possible? The master node wouldn't know about the task id of the policy execution on the node that the policy gets executed on. Only the node where the policy execution actually happens knows this and can return that to the master node. Otherwise we can't report the task id of the actual policy execution.

Reply:
This makes sense to me. Waiting on the task result sounds like a good solution to me then!